spring-beans/src/main/java/org/springframework/beans/BeanUtils.java

```diff
@@ -102,12 +113,11 @@
 		if (clazz.isInterface()) {
 			throw new BeanInstantiationException(clazz, "Specified class is an interface");
 		}
-		try {
-			return instantiateClass(clazz.getDeclaredConstructor());
-		}
-		catch (NoSuchMethodException ex) {
-			throw new BeanInstantiationException(clazz, "No default constructor found", ex);
+		Constructor<T> ctor = findPrimaryConstructor(clazz);
```

Comment:
I guess we'll have to differentiate here: For `instantiateClass(Class)`, we can only deal with the no-arg default constructor (public or non-public) in the regular Java case... not with a non-default unique public constructor that `findPrimaryConstructor` might return to us.

So maybe we'll do some straight `KotlinDelegate.findPrimaryConstructor` vs `clazz.getDeclaredConstructor()` right here in `instantiateClass(Class)`, and only use the public `findPrimaryConstructor` variant for data binding purposes.